libctru/source/services/iruser.c

```diff
@@ -0,0 +1,569 @@
+    Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+
+	cmdbuf[0] = IPC_MakeHeader(0x18, 6, 2); // 0x00180182
+	cmdbuf[1] = iruserSharedMemSize;
+	cmdbuf[2] = recv_buffer_size;
+	cmdbuf[3] = recv_packet_count;
+	cmdbuf[4] = send_buffer_size;
+	cmdbuf[5] = send_packet_count;
+	cmdbuf[6] = bitrate;
+	cmdbuf[7] = IPC_Desc_SharedHandles(1);
+	cmdbuf[8] = iruserSharedMemHandle;
+
+	if(R_FAILED(ret = svcSendSyncRequest(iruserHandle)))return ret;
+	ret = (Result)cmdbuf[1];
+
+	return ret;
+}
+
+Result IRUSER_ReleaseReceivedData(u32 packet_count) {
+    Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+
+	cmdbuf[0] = IPC_MakeHeader(0x19, 1, 0); // 0x00190040
+	cmdbuf[1] = packet_count;
+
+	if(R_FAILED(ret = svcSendSyncRequest(iruserHandle)))return ret;
+	ret = (Result)cmdbuf[1];
+
+	return ret;
+}
+
+Result IRUSER_SetOwnMachineId(u8 id) {
+    Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+
+	cmdbuf[0] = IPC_MakeHeader(0x1A, 1, 0); // 0x001A0040
+	cmdbuf[1] = id;
+
+	if(R_FAILED(ret = svcSendSyncRequest(iruserHandle)))return ret;
+	ret = (Result)cmdbuf[1];
+
+	return ret;
+}
+
+/// This will let you directly read the ir:USER shared memory via a callback.
+void iruserProcessSharedMemory(void(*process_fn)(u8*)) {
+    process_fn((u8*)iruserSharedMem);
+}
+
+/// Read and parse the ir:USER service status data from shared memory.
+IRUSER_StatusInfo iruserGetStatusInfo() {
+    void* shared_mem = iruserSharedMem;
+    IRUSER_StatusInfo status_info;
+    // copy over data
+    memcpy(&status_info, shared_mem, sizeof(status_info));
+
+    return status_info;
+}
+
+Result iruserGetCirclePadProState(IRUSER_Packet* packet, circlePadProInputResponse* response) {
+    if (!packet) goto failure;
+    if (!packet->payload) goto failure;
+    if (packet->payload_length != 6) goto failure;
+    if (packet->payload[0] != CIRCLE_PAD_PRO_INPUT_RESPONSE_PACKET_ID) goto failure;
+    response->cstick.csPos.dx = (u16)(packet->payload[1] | ((packet->payload[2] & 0x0F) << 8));
+    response->cstick.csPos.dy = (u16)(((packet->payload[2] & 0xF0) >> 4) | ((packet->payload[3]) << 4));
+    response->status_raw = packet->payload[4];
+    response->unknown_field = packet->payload[5];
+    return 0;
+    failure: 
+    memset(response, 0, sizeof(circlePadProInputResponse));
+    return -1;
+}
+
+Result iruserCirclePadProCStickRead(circlePosition *pos) {
+    // Result ret = 0;
+    u32 n = 0;
+    IRUSER_Packet* packet = iruserGetPackets(&n);
+    if (n == 0) return -1;
+    if (packet[n-1].payload_length != 6) return RS_INVALIDSTATE;
+    if (packet[n-1].payload[0] != CIRCLE_PAD_PRO_INPUT_RESPONSE_PACKET_ID) return RS_INVALIDSTATE;
+    pos->dx = (s16)(packet[n-1].payload[1] | ((packet[n-1].payload[2] & 0x0F) << 8));
+    pos->dy = (s16)(((packet[n-1].payload[2] & 0xF0) >> 4) | ((packet[n-1].payload[3]) << 4));
+
+    free(packet[n-1].payload);
```

Comment:
Many memory leaks in this function (leaks all n > 1, and the early return also leak)